### ISSUE TITLE
[Feature]: Allow jobs to specify the path to the shell. By default, use `bin/bash`

### ISSUE THREAD

**Issue by charlienewey** [dstackai/dstack#247]
### Problem

Currently `dstack` seems to assume the presence of `bash` in Docker containers, which is not necessarily the case. It would be nice to be able to specify a custom entrypoint (or use the default one provided in the container).

Here's a minimal working example.

```
workflows:
  - name: my-awesome-workflow
    provider: docker
    image: docker:23.0.1-dind
    commands:
      - echo "hooray"
```

This produces the following output;

<pre>(dstack-testing-py3.8) <font color="#4E9A06"><b>➜  </b></font><font color="#06989A"><b>dstack_testing</b></font> <font color="#3465A4"><b>git:(</b></font><font color="#CC0000"><b>master</b></font><font color="#3465A4"><b>) </b></font><font color="#C4A000"><b>✗</b></font> dstack run my-awesome-workflow
<b> RUN              WORKFLOW          SUBMITTED  OWNER          STATUS     TAG  BACKENDS </b>
<b> </b><font color="#C4A000"><b>dull-kangaroo-1</b></font><b> </b><font color="#949494"> </font><font color="#C4A000">my-awesome-work…</font><font color="#949494">  </font><font color="#C4A000">now</font><font color="#949494">        </font><font color="#C4A000">Charles Newey</font><font color="#949494"> </font> <font color="#C4A000">Submitted</font> <font color="#C4A000"><b>     </b></font><font color="#4E9A06"><b> </b></font><font color="#C4A000"><b>local</b></font><font color="#4E9A06"><b>    </b></font>

Provisioning<font color="#C4A000">...</font> It may take up to a minute. <font color="#4E9A06">✓</font>

<font color="#949494">To interrupt, press Ctrl+C.</font>

/usr/local/bin/docker-entrypoint.sh: exec: line 61: /bin/bash: not found
</pre>

Which leads me to believe that `dstack` is overriding the default entrypoint.

### Solution

e.g. add an `entrypoint:` configuration key supported by the `docker` provider.

### Benefit

Allows use of a greater number of `docker` images and increases flexibility.

### Alternatives

None that I'm aware of.

### Would you like to help contributing this feature?

Yes

**Comment by peterschmidt85** [comment]
Here's a suggestion on how to implement it.

1. Add the optional property `shell` to the `docker` provider. Here's an example:

```
workflows:
  - name: my-awesome-workflow
    provider: docker
    image: docker:23.0.1-dind
    shell: /bin/sh
    commands:
      - echo "hooray"
```

This property can be used to override the path to the default shell.

> **Note**
> This means the `Job` class will get a new property `shell` that can be passed by any `Provider`.

2. If `shell` is not defined for a job, use `/bin/sh`.

3. In the `code`, `notebook`, `lab`, and `bash` providers, pass `/bin/bash` to jobs' `shell` property.

**Comment by Egor-S** [comment]
Solved by #253
